Add the support for string reflections on ElementInternals

rniwa opened the pull request:
#### 341101190ec82f392260c91d397d571b97661621
<pre>
Add the support for string reflections on ElementInternals
<a href="https://bugs.webkit.org/show_bug.cgi?id=245028">https://bugs.webkit.org/show_bug.cgi?id=245028</a>

Reviewed by Chris Fleizach.

Added the rest of string reflections on ElementInternals.
Also use Reflect IDL attribute on ElementInternals to share more code between attributes.

* LayoutTests/accessibility/custom-elements/autocomplete-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/autocomplete.html: Added.
* LayoutTests/accessibility/custom-elements/current-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/current.html: Added.
* LayoutTests/accessibility/custom-elements/heading-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/heading.html: Added.
* LayoutTests/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/hidden-button.html: Added.
* LayoutTests/accessibility/custom-elements/liveregions-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/liveregions.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/menuitem.html: Added.
* LayoutTests/accessibility/custom-elements/modal-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/modal.html: Added.
* LayoutTests/accessibility/custom-elements/multiselectable-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/multiselectable.html: Added.
* LayoutTests/accessibility/custom-elements/orientation-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/orientation.html: Added.
* LayoutTests/accessibility/custom-elements/posinset-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/posinset.html: Added.
* LayoutTests/accessibility/custom-elements/pressed-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/pressed.html: Added.
* LayoutTests/accessibility/custom-elements/role-expected.txt:
* LayoutTests/accessibility/custom-elements/role.html:
* LayoutTests/accessibility/custom-elements/slider-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/slider.html: Added.
* LayoutTests/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/table.html: Added.
* LayoutTests/accessibility/custom-elements/textbox-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/textbox.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt: Rebaselined.
* LayoutTests/platform/gtk/accessibility/custom-elements/autocomplete-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/modal-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/pressed-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/textbox-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/current-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/liveregions-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isModalElement const): Added a fallback for default ARIA
(WebCore::nodeHasRole): Ditto.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::hasAttribute const): Ditto.
(WebCore::AccessibilityObject::getAttribute const):

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::CustomElementDefaultARIA::setValueForAttribute): Use QualifiedName instead of
AtomString for the mapping attributes.
(WebCore::CustomElementDefaultARIA::valueForAttribute const): Ditto.
(WebCore::CustomElementDefaultARIA::hasAttribute const): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.h:
(WebCore::CustomElementDefaultARIA):

* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::setRole): Deleted.
(WebCore::ElementInternals::role): Deleted.
(WebCore::ElementInternals::setAriaRoleDescription): Deleted.
(WebCore::ElementInternals::ariaRoleDescription): Deleted.
(WebCore::ElementInternals::setAriaLabel): Deleted.
(WebCore::ElementInternals::ariaLabel): Deleted.
(WebCore::ElementInternals::setAriaValueForAttribute): Deleted.
(WebCore::ElementInternals::ariaValueForAttribute): Deleted.
(WebCore::ElementInternals::setAttributeWithoutSynchronization): Notify AX object cache.
(WebCore::ElementInternals::attributeWithoutSynchronization const):

* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:

* Source/WebCore/dom/QualifiedName.h: Added the missing include for AtomStringHash.h.

Canonical link: <a href="https://commits.webkit.org/254549@main">https://commits.webkit.org/254549@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3caedcd046d10330b05996ec36b47143f3b65aa1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20114 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98677 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154980 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32401 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27921 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81725 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93097 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25738 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76277 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25704 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80615 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80555 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68688 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30174 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/14618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29902 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15566 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33350 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38523 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1335 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32054 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34657 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mrego reviewed (commented):
Looks good to me, but it'd be nice someone with more experience on `Source/WebCore/accessibility/` take a look.

twilco commented on LayoutTests/accessibility/custom-elements/autocomplete.html:
```diff
@@ -0,0 +1,36 @@
+
+</script>
+
+<script src="../../../resources/js-test-post.js"></script>
```
I think the  `js-test-post.js` includes are not necessary with the `js-test.js` include above. Ditto for the other tests.

rniwa replied:
Fixed.

twilco commented on LayoutTests/accessibility/custom-elements/autocomplete.html:
```diff
@@ -0,0 +1,36 @@
+<custom-list id="custom-list"></custom-checkbox>
+<custom-list id="custom-textbox" role="textbox" aria-autocomplete="inline"></custom-checkbox>
+<p id="description"></p>
+<div id="console"></div>
```
I think:

```
<p id="description"></p>
<div id="console"></div>
 ```
 
 Is not necessary when including `js-test.js`. Ditto for the other tests.

rniwa replied:
Fixed.

twilco commented on LayoutTests/accessibility/custom-elements/modal.html:
```diff
@@ -0,0 +1,41 @@
+        shouldBeTrue('!accessibilityController.accessibleElementById("background-content") || accessibilityController.accessibleElementById("background-content").isIgnored');
+        document.getElementById('custom-dialog').setAttribute('aria-modal', 'false');
+        await new Promise(requestAnimationFrame);
+        shouldBeFalse('accessibilityController.accessibleElementById("background-content").isIgnored');
```
`accessibility-helpers.js` defines an `expectAsyncExpression` function that might make this test more explicit than awaiting RAF:

```
document.getElementById('custom-dialog').setAttribute('aria-modal', 'false');
await expectAsyncExpression('accessibilityController.accessibleElementById("background-content").isIgnored', 'false');
```

rniwa replied:
Neat. Done.

twilco reviewed (commented):
Thanks for doing this!

twilco commented on Source/WebCore/dom/ElementInternals.cpp:
```diff
@@ -45,47 +46,21 @@ ShadowRoot* ElementInternals::shadowRoot() const
-{
-    m_element->customElementDefaultARIA().setValueForAttribute(key, value);
+    if (AXObjectCache* cache = element->document().existingAXObjectCache())
+        cache->deferAttributeChangeIfNeeded(name, element.get());
```
Unfortunately, a commit landed a few days ago that will require a rebase and change on this line.

https://github.com/WebKit/WebKit/commit/a313324db71fdc82a9ecfd4071e1569b1029281a

The signature of this function is now `AXObjectCache::deferAttributeChangeIfNeeded(Element* element, const QualifiedName& attrName, const AtomString& oldValue, const AtomString& newValue)`

rniwa replied:
Fixed.

fleizach reviewed (approved):
Thanks!

rniwa committed:
Add the support for string reflections on ElementInternals
https://bugs.webkit.org/show_bug.cgi?id=245028

Reviewed by Chris Fleizach.

Added the rest of string reflections on ElementInternals.
Also use Reflect IDL attribute on ElementInternals to share more code between attributes.

* LayoutTests/accessibility/custom-elements/autocomplete-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/autocomplete.html: Added.
* LayoutTests/accessibility/custom-elements/current-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/current.html: Added.
* LayoutTests/accessibility/custom-elements/heading-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/heading.html: Added.
* LayoutTests/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/hidden-button.html: Added.
* LayoutTests/accessibility/custom-elements/liveregions-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/liveregions.html: Added.
* LayoutTests/accessibility/custom-elements/menuitem-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/menuitem.html: Added.
* LayoutTests/accessibility/custom-elements/modal-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/modal.html: Added.
* LayoutTests/accessibility/custom-elements/multiselectable-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/multiselectable.html: Added.
* LayoutTests/accessibility/custom-elements/orientation-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/orientation.html: Added.
* LayoutTests/accessibility/custom-elements/posinset-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/posinset.html: Added.
* LayoutTests/accessibility/custom-elements/pressed-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/pressed.html: Added.
* LayoutTests/accessibility/custom-elements/role-expected.txt:
* LayoutTests/accessibility/custom-elements/role.html:
* LayoutTests/accessibility/custom-elements/slider-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/slider.html: Added.
* LayoutTests/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/table.html: Added.
* LayoutTests/accessibility/custom-elements/textbox-expected.txt: Added.
* LayoutTests/accessibility/custom-elements/textbox.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/form-associated/ElementInternals-accessibility-expected.txt: Rebaselined.
* LayoutTests/platform/gtk/accessibility/custom-elements/autocomplete-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/modal-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/pressed-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/platform/gtk/accessibility/custom-elements/textbox-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/current-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/hidden-button-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/liveregions-expected.txt: Added.
* LayoutTests/platform/mac-wk1/accessibility/custom-elements/table-expected.txt: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::isModalElement const): Added a fallback for default ARIA
(WebCore::nodeHasRole): Ditto.

* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::hasAttribute const): Ditto.
(WebCore::AccessibilityObject::getAttribute const):

* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::CustomElementDefaultARIA::setValueForAttribute): Use QualifiedName instead of
AtomString for the mapping attributes.
(WebCore::CustomElementDefaultARIA::valueForAttribute const): Ditto.
(WebCore::CustomElementDefaultARIA::hasAttribute const): Added.

* Source/WebCore/dom/CustomElementDefaultARIA.h:
(WebCore::CustomElementDefaultARIA):

* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::setRole): Deleted.
(WebCore::ElementInternals::role): Deleted.
(WebCore::ElementInternals::setAriaRoleDescription): Deleted.
(WebCore::ElementInternals::ariaRoleDescription): Deleted.
(WebCore::ElementInternals::setAriaLabel): Deleted.
(WebCore::ElementInternals::ariaLabel): Deleted.
(WebCore::ElementInternals::setAriaValueForAttribute): Deleted.
(WebCore::ElementInternals::ariaValueForAttribute): Deleted.
(WebCore::ElementInternals::setAttributeWithoutSynchronization): Notify AX object cache.
(WebCore::ElementInternals::attributeWithoutSynchronization const):

* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:

* Source/WebCore/dom/QualifiedName.h: Added the missing include for AtomStringHash.h.

Canonical link: https://commits.webkit.org/254549@main
